ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/Genesis/Tests/Uniform.hs

```diff
@@ -0,0 +1,111 @@
+{-# LANGUAGE BlockArguments      #-}
+{-# LANGUAGE DerivingStrategies  #-}
+{-# LANGUAGE LambdaCase          #-}
+{-# LANGUAGE NamedFieldPuns      #-}
+{-# LANGUAGE RecordWildCards     #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TupleSections       #-}
+{-# LANGUAGE TypeFamilies        #-}
+{-# LANGUAGE TypeOperators       #-}
+
+module Test.Consensus.Genesis.Tests.Uniform (tests) where
+
+import           Control.Monad.IOSim (runSimOrThrow)
+import           Data.List (intercalate)
+import qualified Data.List.NonEmpty as NonEmpty
+import           Data.Word (Word64)
+import           Ouroboros.Consensus.Block (ChainHash (..), HeaderHash)
+import           Ouroboros.Consensus.Config (SecurityParam (SecurityParam))
+import           Ouroboros.Consensus.Util.Condense (condense)
+import qualified Ouroboros.Network.AnchoredFragment as AF
+import           System.Random.Stateful (runSTGen_)
+import           Test.Consensus.BlockTree (btTrunk)
+import           Test.Consensus.Genesis.Setup
+import           Test.Consensus.Genesis.Setup.Classifiers
+import           Test.Consensus.PeerSimulator.Run
+                     (SchedulerConfig (scEnableGdd), noTimeoutsSchedulerConfig)
+import           Test.Consensus.PeerSimulator.StateView
+import           Test.Consensus.PointSchedule
+import qualified Test.QuickCheck as QC
+import           Test.QuickCheck
+import           Test.QuickCheck.Extras (unsafeMapSuchThatJust)
+import           Test.QuickCheck.Random (QCGen)
+import           Test.Tasty
+import           Test.Tasty.QuickCheck
+import           Test.Util.Orphans.IOLike ()
+import           Test.Util.TestBlock (TestHash (TestHash))
+import           Test.Util.TestEnv (adjustQuickCheckTests)
+
+tests :: TestTree
+tests =
+  adjustQuickCheckTests (`div` 10) $
+  testProperty "uniform random schedule" prop_uniformSchedule
+
+genChainsAndSchedule :: PointScheduleConfig -> ScheduleType -> QC.Gen (GenesisTest, PointSchedule)
+genChainsAndSchedule scheduleConfig scheduleType =
+  unsafeMapSuchThatJust do
+    numAdversaries <- QC.choose (1, 4)
+    gt <- genChains numAdversaries
+    seed :: QCGen <- arbitrary
+    pure ((gt,) <$> runSTGen_ seed (\ g -> genSchedule g scheduleConfig scheduleType (gtBlockTree gt)))
+
+simpleHash ::
+  HeaderHash block ~ TestHash =>
+  ChainHash block ->
+  [Word64]
+simpleHash = \case
+  GenesisHash -> []
+  BlockHash (TestHash h) -> reverse (NonEmpty.toList h)
+
+makeProperty ::
+  GenesisTest ->
+  Int ->
+  StateView ->
+  [PeerId] ->
+  Property
+makeProperty genesisTest advCount StateView {svSelectedChain} killed =
+  classify genesisWindowAfterIntersection "Full genesis window after intersection" $
+  classify (isOrigin immutableTip) "Immutable tip is Origin" $
+  classify (length killed > 0) "Some peers were killed" $
+  classify (length killed == advCount) "All adversaries were killed" $
+  counterexample killedPeers $
+  conjoin [
+    property (isHonest immutableTip),
```

Comment:
```suggestion
    counterexample ("The immutable tip is honest: " ++ show immutableTip) $
      property (isHonest immutableTip),
```

Reply:
That's only printed when the property fails (not sure about the conjunction, but I'd assume it is locally scoped)

Reply:
Indeed. The msg is to hint more explicitly which property has failed, and to print the immutableTip in that case.

Reply:
should say "is not honest" then though 😅

Reply:
✅